include/cadmium/celldevs/cell/cell.hpp

```diff
@@ -0,0 +1,224 @@
+            buffer->add_to_buffer(initial_state, T());  // At t = 0, every cell communicates its state to its neighbors
+        }
+
+        /************** USER-DEFINED METHODS **************/
+        /// @return cell's next state.
+        virtual S local_computation() const { return state.current_state; }
+        /// @return delay to be applied before communicating to neighbors a new state.
+        virtual T output_delay(S const &cell_state) const { return std::numeric_limits<T>::infinity(); }
+
+        /****************** DEVS METHODS ******************/
+        /// @brief the internal transition function clears output delayer buffer and updates clock and next time advance.
+        void internal_transition() {
+            buffer->pop_buffer();
+            clock += time_advance();
+            next_internal = buffer->next_timeout() - clock;
+        }
+
+        /**
+         * External transition function.
+         * @brief It updates clock and next internal event. Then, it refreshes neighbors' state and computes next cell state.
+         *        if the new cell state is different to the current state, it adds the new state to the output delayer buffer.
+         * @param e elapsed time from the last event.
+         * @param mbs message bag containing new neighbors' state messages.
+         */
+        void external_transition(T e, typename cadmium::make_message_bags<input_ports>::type mbs) {
+            // Update clock and next internal event
+            clock += e;
+            next_internal -= e;
+            // Refresh the neighbors' current state
+            std::vector<cell_state_message<C, S>> bagPortIn = cadmium::get_messages<typename cell_ports_def<C, S>::cell_in>(mbs);
+            for (cell_state_message<C, S> msg: bagPortIn) {
+                auto it = state.neighbors_state.find(msg.cell_id);
+                if (it != state.neighbors_state.end()) {
+                    state.neighbors_state[msg.cell_id] = msg.state;
+                }
+            }
+            // Compute next state
+            S next = local_computation();
+            // If next state is not the current state, then I change my state and schedule my next internal transition
+            if (next != state.current_state) {
+                buffer->add_to_buffer(next, clock + output_delay(next));
+                next_internal = buffer->next_timeout() - clock;
+            }
+            state.current_state = next;
+        }
+
+        /// Confluence transition function.
+        void confluence_transition(T e, typename cadmium::make_message_bags<input_ports>::type mbs) {
+            internal_transition();
+            external_transition(T(), move(mbs));
+        }
+
+        /// Time advance function.
+        T time_advance() const { return next_internal; }
+
+        /// @return the next message to be transmitted from the output delayer buffer
+        typename cadmium::make_message_bags<output_ports>::type output() const {
```

Comment:
Shouldm't be provided by user?
I'm really confuse about this class, you providing a lot of things that I would expect user to provide, can you comment on what is the expectation from the user point of view?

Reply:
The same as above